### DISCUSSION TITLE
enable sfdx force:source:deploy to accept test suites

### DISCUSSION THREAD

**Discussion by AllanOricil** [forcedotcom/cli#2349] [Ideas]
**Is your feature request related to a problem? Please describe.**

Im trying to run test suites when running sfdx force:source:deploy. I can do it by myself querying ApexTestSuite and TestSuiteMembership to generate a list of apex test classes, but it would nice to have this available by default in sfdx. 

At the moment only `sfdx force:apex:test:run` supports a list of test suites. This command has a flag called `-s | --suite-names` which accepts `Comma-separated list of Apex test suite names to run; default is all suites`

**What are you trying to do**
Execute tests using test suites when running `sfdx force:source:deploy`

**Describe the solution you'd like**

`sfdx force:source:deploy --test-suites test_suite_1,test_suite_2,test_suite_3`

**Describe alternatives you've considered**
convert test suites to a a list of apex tests classes, and then pass this list to the `-r` flag in `sfdx force:source:deploy`

**Additional context**
N/A

[a comment by an automated account is not shown]

**Comment by mitchspano** [comment]
I see that this issue was closed, but when I try to use the command as proposed, it does not work. Has this been added to the CLI?

```
sfdx force:source:deploy -p "force-app/main/default/aura/Foo" --test-suite MySuite
```
```
ERROR running force:source:deploy:  Unexpected arguments: --test-suite, MySuite
See more help with --help
```

**Comment by peternhale** [comment]
Should not of be closed. The work item has yet to be scheduled.

**Comment by mitchspano** [comment]
Thanks for the quick response @peternhale! I am also curious if this will be able to be used in conjunction with `-l RunSpecifiedTests` or if they will be mutually exclusive. I would hope to be able to run specified tests as well as the elements of a test suite - something like:
```
sfdx force:source:deploy -p "force-app/main/default/aura/Foo"  -l RunSpecifiedTests -r FooControllerTest --test-suite MySuite
```

**Comment by peternhale** [comment]
I do not know at this moment, but I will make sure your comment is included in the work item.

**Comment by steals** [comment]
Subscribing for this feature as it would dramatically improve our deployment time.

**Comment by AllanOricil** [comment]
Relates
 https://github.com/forcedotcom/salesforcedx-vscode/issues/3596

**Comment by mshanemc** [comment]
We don't want to implement this as described (query suites to translate it into test names) because that quickly hits a URL length limit.  What you really want is for the [metadata API](https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/meta_deploy.htm) to accept test suites in its `runTests` property, or to have a separate prop for tests

**Reply by AllanOricil** [reply]
you rock, sir @mshanemc 
didn't know about those limits

**Reply by AllanOricil** [reply]
What if you break the number of requests based on the Max URL length (constant minus a threshold just to be safe) ? @mshanemc 

````js
...
try{
// use something like axios-retry here
 const requests = [array of requests based on url length]
 const results = await Promise.all(requests)
}catch(e){
 ...
}finally{...}
...
````

Could this ~~hack~~ solution work? :D

**Reply by AllanOricil** [reply]
you could also maybe use that sf api that process multiple rest api requests at once, instead of one after another. I remember  I used it once, but Im not sure if it would work here. If this is possible, then you could break that array of requests into a further group of batched requests @mshanemc

**Reply by AllanOricil** [reply]
Forget what I said. I understood the problem now.

**Reply by AllanOricil** [reply]
However, I did not find any limits related to the max number of apex classes you can specify in the runTests array
 https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/meta_deploy_run_specific_tests.htm
